### DISCUSSION TITLE
Interceptor에서 Service 사용

### DISCUSSION THREAD

**Discussion by HyeonbinSa** [woowacourse-teams/2022-ternoko#470] [General]
### As-is
```java
@Component
@RequiredArgsConstructor
public class AuthInterceptor implements HandlerInterceptor {

    private final AuthService authService;

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
        if (CorsUtils.isPreFlightRequest(request)) {
            return true;
        }
        String header = request.getHeader(AUTHORIZATION);
        if (header == null) {
            throw new InvalidTokenException(ExceptionType.UNAUTHORIZED_MEMBER);
        }
        return authService.isValid(header);
    }
}
```
```java
class AuthService {
    ...
    public boolean isValid(final String header) {
        final String token = AuthorizationExtractor.extract(header);
        jwtProvider.validateToken(token);
        return true;
    }
   ....
}
```
현재 AuthInterceptor에서 Service 클래스를 사용하고 있습니다.
### To-be
Interceptor에 대해 학습을 하면서 생각이 든 부분이 Interceptor는 DispatcherServlet과 Handler를 수행하는 Controller 사이에서 동작하는데 Interceptor에서 Service를 가지고 검증하는 것이 맞을까라는 생각을 했습니다.
AuthService의 isVaild() 메서드를 봤을 때 AuthorizationExtractor와 JwtProvider 두 클래스를 사용하는 메서드인만큼 이 메서드를 Interceptor로 옮기는 것은 어떨까 생각합니다.

바꾸게된다면 아래와 같은 코드일 것으로 예상하는데 다른 분들의 의견이 궁금합니다.
```java
@Component
@RequiredArgsConstructor
public class AuthInterceptor implements HandlerInterceptor {

    private final JwtProvider jwtProvider;

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
        if (CorsUtils.isPreFlightRequest(request)) {
            return true;
        }
        String header = request.getHeader(AUTHORIZATION);
        if (header == null) {
            throw new InvalidTokenException(ExceptionType.UNAUTHORIZED_MEMBER);
        }
        return isValid(header);
    }

    private boolean isValid(final String header) {
        final String token = AuthorizationExtractor.extract(header);
        jwtProvider.validateToken(token);
        return true;
    }
}

```

[노션 정리 글 링크](https://necessary-sundial-178.notion.site/Filter-vs-Interceptor-vs-ArgumentResolver-a292b6e88e56483e8df0386f506b262e)

**Comment by HyeonbinSa** [comment]
<img width="523" alt="image" src="https://user-images.githubusercontent.com/36189291/195646322-63ac57c5-765d-447b-8eaa-c599fdb72da5.png">

[블로그](https://steady-coding.tistory.com/601) 를 보니 인터셉터가 서비스 로직을 활용하여 요청 정보 정합성 처리를 담당한다. 라고 설명되어있네요.

**Comment by soominsohn** [comment]
저도 동의합니다! isValid 메서드가 AuthInterceptor에서만 사용되고 있고, AuthInterceptor의 책임인 것 같네요. 
또한` Interceptor가 서비스 로직을 가져도되는가?`에 대해 고민이 됐었는데 바니가 공유해주신 자료를 보니 Interceptor의 역할인 것도 같네요!

**Comment by Juhyung990122** [comment]
isValid 가 인증과 관련된 기능인지를 정한 뒤 사용하시면 좋을 것 같아요~
요청 들어온 토큰이 우리가 발급한 토큰이 맞는지 검증하는 과정이 
서비스 내의 로그인 로그아웃, 회원가입과 같은 auth 기능에 포함이 된다고 생각하신다면 authService
토큰 검증이 사용자를 검증하는 서비스 내의 auth 기능 과는 분리되어있다고 생각하신다면 바니 의견대로 interceptor에 넣는게 맞아보입니다~

**Comment by her0807** [comment]
## AuthService 의 내부 로직
현재 AuthService 에서는 로그인, 로그아웃, 회원가입 auth 기능이 있어요. 
저는 토큰검증 자체가 로그인 기능이기 때문에 service 에 있는게 더 적절한 책임이라고 생각돼요.

## Interceptor 
여기에 isValid 만을 사용하기 위해서 JwtProvider 의존성 주입을 받는게 무거워 보입니다. 🥲

```java

   public boolean isValid(final String header) {
        final String token = AuthorizationExtractor.extract(header);
        jwtProvider.validateToken(token);
        return true;
    }
```
